src/cocaine-app/infrastructure.py

```diff
@@ -116,41 +118,38 @@ def init(self, node, job_finder):
         self.ns_settings = {}
         self._sync_ns_settings()
 
+    def schedule_history_update(self):
+        self.__tq.add_task_in(
+            self.TASK_UPDATE,
+            config.get('infrastructure_update_period', 300),
+            self._update_state)
+
     def _start_tq(self):
         self.__tq.start()
 
     def get_group_history(self, group_id):
-        couples_history = []
-        for couple in self.state[group_id]['couples']:
-            couples_history.append({'couple': couple['couple'],
-                                    'timestamp': couple['timestamp']})
-        nodes_history = []
-        for node_set in self.state[group_id]['nodes']:
-
-            nodes_history.append({
-                'set': node_set['set'],
-                'timestamp': node_set['timestamp'],
-                'type': self.__node_state_type(node_set),
-            })
-        return {'couples': couples_history,
-                'nodes': nodes_history}
+        group_history = self.group_history_finder.group_history(group_id)
+        if group_history is None:
+            raise ValueError('History for group {} is not found'.format(group_id))
+        return group_history
 
     def node_backend_in_last_history_state(self, group_id, host, port, backend_id):
-        if group_id not in self.state:
-            raise ValueError('Group {0} history is not found'.format(group_id))
+        group_history = self.group_history_finder.group_history(group_id)
+        if group_history is None:
+            raise ValueError('History for group {} is not found'.format(group_id))
```

Comment:
Sorry, misread your comment. You're completely right: https://github.com/yandex/mastermind/commit/12754092e0ee686119fe4532c51c91e0d0228acb